tests/unit/packaging/test_models.py

```diff
@@ -266,6 +266,99 @@ def test_acl(self, db_session):
             key=lambda x: x[1],
         )
 
+    def test_acl_for_quarantined_project(self, db_session):
+        """
+        If a Project is quarantined, the Project ACL should disallow any modifications.
+        """
+        project = DBProjectFactory.create(lifecycle_status="quarantine-enter")
+        owner1 = DBRoleFactory.create(project=project)
+        owner2 = DBRoleFactory.create(project=project)
+        maintainer1 = DBRoleFactory.create(project=project, role_name="Maintainer")
+        maintainer2 = DBRoleFactory.create(project=project, role_name="Maintainer")
+
+        organization = DBOrganizationFactory.create()
+        owner3 = DBOrganizationRoleFactory.create(organization=organization)
+        DBOrganizationProjectFactory.create(organization=organization, project=project)
+
+        team = DBTeamFactory.create()
+        owner4 = DBTeamRoleFactory.create(team=team)
+        DBTeamProjectRoleFactory.create(
+            team=team, project=project, role_name=TeamProjectRoleType.Owner
+        )
+
+        publisher = GitHubPublisherFactory.create(projects=[project])
+
+        acls = []
+        for location in lineage(project):
+            try:
+                acl = location.__acl__
+            except AttributeError:
+                continue
+
+            if acl and callable(acl):
+                acl = acl()
+
+            acls.extend(acl)
+
+        assert acls == [
+            (
+                Allow,
+                "group:admins",
+                (
+                    Permissions.AdminDashboardSidebarRead,
+                    Permissions.AdminObservationsRead,
+                    Permissions.AdminObservationsWrite,
+                    Permissions.AdminProhibitedProjectsWrite,
+                    Permissions.AdminProjectsDelete,
+                    Permissions.AdminProjectsRead,
+                    Permissions.AdminProjectsSetLimit,
+                    Permissions.AdminProjectsWrite,
+                    Permissions.AdminRoleAdd,
+                    Permissions.AdminRoleDelete,
+                ),
+            ),
+            (
+                Allow,
+                "group:moderators",
+                (
+                    Permissions.AdminDashboardSidebarRead,
+                    Permissions.AdminObservationsRead,
+                    Permissions.AdminObservationsWrite,
+                    Permissions.AdminProjectsRead,
+                    Permissions.AdminProjectsSetLimit,
+                    Permissions.AdminRoleAdd,
+                    Permissions.AdminRoleDelete,
+                ),
+            ),
+            (
+                Allow,
+                "group:observers",
+                Permissions.APIObservationsAdd,
+            ),
+            (
+                Allow,
+                Authenticated,
+                Permissions.SubmitMalwareObservation,
+            ),
+        ] + sorted(
+            [(Allow, f"oidc:{publisher.id}", [Permissions.ProjectsUpload])],
```

Comment:
I'm reading this as allowing uploads via Trusted Publisher can still occur for quarantined projects, while uploads via API token would 403. Is that correct?

Is there a reason we would continue to allow quarantined projects to upload one way but not another?

Should we disallow Trusted Publisher uploads, or continue to allow API Token uploads to maintain consistency?

Reply:
Thinking a moment more, I think we probably want to continue to allow uploads either way. In the event of a false-positive quarantine, breaking uploads is generally more disruptive to projects.

Reply:
> I'm reading this as allowing uploads via Trusted Publisher can still occur for quarantined projects, while uploads via API token would 403. Is that correct?

That is correct, since the change is for the user token ACL to disallow modifications.

> Thinking a moment more, I think we probably want to continue to allow uploads either way. In the event of a false-positive quarantine, breaking uploads is generally more disruptive to projects.

I agree that disallowing uploads is more disruptive, but if the project is using Trusted Publishing, then it's not prevented from uploads. I don't think I've seen any malware **yet** come through a TP upload.
I considered disallowing TP as well, and decided that if they went to the trouble of creating, linking a TP, then at least we'd have a backlink to a more concrete location from whence the code came from.

If you feel strongly that a user-scoped token should continue to allow uploads to a quarantine-flagged project (and not other projects), then I'll modify to allow that, but I wanted to provide the "press pause on this project until it's been reviewed" semantics.
The risk of allowing further uploads is that we have more junk in storages vs not.

Reply:
I am still a slight lean towards "allow all uploads to continue", but I am a strong lean towards "the constraint should be consistent between projects".

Reply:
In that light, I agree for consistency, so if allowing TP, I'll update to allow user-based uploads as well.
We can shift that policy later if we find that the support burden rises if folks are uploading new releases and reach out to determine why they are not allowed to install them.